lib/rubocop/cop/rails/migration_timestamp.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+require 'time'
+
+module RuboCop
+  module Cop
+    module Rails
+      # Checks that migration file names start with a valid timestamp.
+      #
+      # @example
+      #   # bad
+      #   # db/migrate/bad.rb
+
+      #   # bad
+      #   # db/migrate/123_bad.rb
+
+      #   # bad
+      #   # db/migrate/20171301000000_bad.rb
+      #
+      #   # good
+      #   # db/migrate/20170101000000_good.rb
+      #
+      class MigrationTimestamp < Base
+        include RangeHelp
+
+        MSG = 'Migration file name must start with `YYYYmmddHHMMSS_` timestamp.'
+
+        def on_new_investigation
+          file_path = processed_source.file_path
+          return unless file_path.include?('db/migrate')
+
+          timestamp = File.basename(file_path)[/\A\d{14}(?=_)/]
```

Comment:
I was about to suggest splitting at the first `_` for a more readable version, but it turns out `Time.strptime` will accept more characters than the pattern needs, and non-numeric characters too...


```rb
# Proper time
Time.strptime("20230505163015", "%Y%m%d%H%M%S")
#=> 2023-05-05 16:30:15 +0900

# Extra character
Time.strptime("20230505163015a", "%Y%m%d%H%M%S")
#=> 2023-05-05 16:30:15 +0900

# Non-numeric character
irb(main):008:0> Time.strptime("2023050516301a", "%Y%m%d%H%M%S")
=> 2023-05-05 16:30:01 +0900
```

Regex looks good! 👍🏻